### PR TITLE
messages app

### PR DESCRIPTION
**Pull request by brucexhan**

### CODE REVIEW

File: apps/wolk.js [hunk 1]
```diff
@@ -260,13 +260,13 @@ class WOLK {
...
         await window.crypto.subtle.generateKey({
-            name: "RSASSA-PKCS1-v1_5",
+            name: "RSA-OAEP",
             modulusLength: 2048,
             publicExponent: new Uint8Array([0x01, 0x00, 0x01]),
             hash: {
                 name: "SHA-256"
             }
-        }, true, ["sign", "verify"]).then((rsaKey) => {
+        }, true, ["encrypt", "decrypt"]).then((rsaKey) => {
```
Comment (brucexhan) [inline apps/wolk.js]: generate encrypt/decrypt rsa keys for messaging

File: apps/wolk.js [hunk 2]
```diff
@@ -953,6 +955,24 @@ function uint8ArrayToHex(bytes) {
...
+        var stringValue = value.toString(16)
+        // We use concatenation and slice for padding
+        var padding = '00000000'
+        var paddedValue = (padding + stringValue).slice(-padding.length)
+        hexCodes.push(paddedValue);
+    }
+
+    // Join all the hex strings into one
+    return hexCodes.join("");
+}
```
Comment (brucexhan) [inline apps/wolk.js]: another buffer-to-hex function for converting encrypted friendsKey bytes to hex string

File: apps/messages/home.html [hunk 3]
```diff
@@ -0,0 +1,1005 @@
...
+        { owner: owner, bucket: ActionBucket, key: friendReqKey, value: reqDetails, hdrs: {} },
+        { owner: owner, bucket: ActionBucket, key: friendReqKey, value: actionMeta, hdrs: { "BodyType": "metadata" } }
+    ];
+
+    var txnHashes = await wolkjs.setKeys(txnArray);
+    console.log(txnHashes);
+}
+*/
+
+async function sendFriendReq(e) {
```
Comment (brucexhan) [inline apps/messages/home.html]: sendFriendReq function is used when user find a stranger on his friend's friend list

File: apps/messages/home.html [hunk 4]
```diff
@@ -0,0 +1,1005 @@
...
+
+    var recipientRsaPublicKey = await window.crypto.subtle.importKey("jwk", JSON.parse(recipientRsaPublicKeyJSON), { name: "RSA-OAEP", hash: { name: "SHA-256" } }, false, ["encrypt"]);
+
+    var encryptedfriendsKeyBytes = await window.crypto.subtle.encrypt({ name: "RSA-OAEP" }, recipientRsaPublicKey, friendsKeyBytes);
+
+    console.log(encryptedfriendsKeyBytes);
+    console.log(arbuf2hex(encryptedfriendsKeyBytes));
+
+    return arbuf2hex(encryptedfriendsKeyBytes);
+}
```
Comment (brucexhan) [inline apps/messages/home.html]: genEncryptedFriendsKey function is called inside acceptReq function when a user sends a BefriendAction.
firendsKey string before encryption is constructed as ${agent/owner}_${recipient}_${agent-unique-salt}. An example: user01 sends BefriendAction to user02, firendsKey string will be "user01_user02_asdfg". The `asdfg` at the end is a few characters from user01's rsa private key. then we can assume the firendsKey before encryption is deterministic and also hard to be hacked. 

In order to store the encryptedfriendsKey to Action table, encryptedfriendsKeyBytes needs to be converted to hex string.

File: apps/messages/home.html [hunk 5]
```diff
@@ -0,0 +1,1005 @@
...
+            var friendsKeyBytes = await digestMessage(firendsKeyStr);
+            console.log(friendsKeyBytes);
+            var friendsKeyObj = await window.crypto.subtle.importKey("raw", friendsKeyBytes, "AES-GCM", false, ["encrypt", "decrypt"]);
+
+            console.log(friendsKeyObj)
+            return friendsKeyObj;
+        }
+
+    });
+}
```
Comment (brucexhan) [inline apps/messages/home.html]: getFriendKey function explanation:
scenario: user01 sends BefriendAction to user02, friendsKey between user01 and user02 is genreated on user01's side. so when user01 talks to user02, user01 retrieves friendsKey before encryption by recontructing raw friendsKey by following the formula `user01_user02_fewCharsFromUser01RSAPrivateKey`. when user02 talks to user01, user02 retrieves friendsKey by scanning Action table to look for the BefriendAction data between user02 and user01 and then decrypt the encrypted friendsKey using user02's rsa private key.

File: apps/messages/home.html [hunk 6]
```diff
@@ -0,0 +1,1005 @@
...
+    var encyptedMsgByte = hexStringToByte(msg);
+
+    var msgByte = await window.crypto.subtle.decrypt({ name: "AES-GCM", iv: iv }, friendKey, encyptedMsgByte);
+
+    var dec = new TextDecoder();
+
+    return dec.decode(msgByte);
+}
+
+async function loadMessages() {
```
Comment (brucexhan) [inline apps/messages/home.html]: loadMessages: scan Action table to find all the MessageAction rows between the two users, then sort the messages in timestamp order, then divide them into come-in messages and sent-out messages based on the agent and recipient fields in the data rows.

File: apps/messages/navbar.html [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        <ul class="navbar-nav">
+            <li class="nav-item">
+                <!-- <a class="nav-link" href="home.html">Home</a> -->
+            </li>
+        </ul>
+        <form class="form-inline search-bar">
+            <input class="form-control mr-sm-2" type="search" placeholder="Send friend request by name" aria-label="Send" id="sendFriendReqInput">
+            <button class="btn btn-info" type="button" id="sendFriendReqBtn">Send</button>
+            &nbsp;<span class="err" id="sendFriendReqErrSpan"></span>
+        </form>
```
Comment (brucexhan) [inline apps/messages/navbar.html]: Add a temporary form on the navigation bar to send AskAction (send a friend request to an existing account)

File: apps/messages/profile.html [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+
+function displaySelectedFile(e) {
+    // console.log(e.id);
+    var spanId = e.id + "-span";
+
+    document.getElementById(spanId).innerHTML = document.getElementById(e.id).files[0].name;
+}
+
+
+async function update() {
```
Comment (brucexhan) [inline apps/messages/profile.html]: update profile: If user already has a profile image in Person table, there will be an wolk url for the profile image in the avatar section. Then when the user tries to update his/her profile, he/she can choose a new image from local disk (we will upload the new image to server and have new transactions for ImageObject table) or keep the old profle image (we will read the old profile image wolk url and carry it to the new Person table transaction and there will be no new transaction for ImageObject table)

File: apps/messages/messages.js [hunk 9]
```diff
@@ -0,0 +1,440 @@
...
+            for (var i = 0; i < profiles.length; i++) {
+                profileValObj[profiles[i].timestamp] = profiles[i];
+                profileTSArr.push(profiles[i].timestamp);
+                profileTSArr.sort();
+            }
+
+            return callback(null, profileValObj[profileTSArr[profileTSArr.length - 1]]);
+        }
+    });
+}
```
Comment (brucexhan) [inline apps/messages/messages.js]: getLatestProfile: scan the Person table to find all the data rows from a specific user, then sort all the rows with timestamp, find the latest row, and return it.

File: apps/messages/messages.js [hunk 10]
```diff
@@ -0,0 +1,440 @@
...
+
+function getLatestProfilePromise(accountName) {
```
Comment (brucexhan) [inline apps/messages/messages.js]: getLatestProfilePromise is a return-promise version of getLatestProfile

File: apps/messages/messages.js [hunk 11]
```diff
@@ -0,0 +1,440 @@
...
+                    profileTSArr.push(profiles[i].timestamp);
+                    profileTSArr.sort();
+                }
+                resolve(profileValObj[profileTSArr[profileTSArr.length - 1]]);
+            }
+        });
+    });
+}
+
+async function sendFriendReqNavBar(recipient, callback) {
```
Comment (brucexhan) [inline apps/messages/messages.js]: sendFriendReqNavBar is used for the send friend request on navigation bar. 
It first checks if the input account name (the friend request recipient) is existing on the chain and if it's not the owner's name (one cannot send friend request to himself/herself), then send out the transaction if input account name is valid.

File: apps/messages/profile.html (hunk 8 above)
Comment (sourabhniyogi) [reply apps/messages/profile.html]: We probably need to do a PATCH operation here.

**Pull request merged**